### ISSUE TITLE
EXGBoost crashes when dataset is too big

### ISSUE THREAD

**Issue by darkblueorange** [acalejos/exgboost#23]
First, thanks a lot for this NIF!

On Livebook as in plain Elixir module, :exgboost NIF crashes when dataset size exceeds 208_600 lines. 

Created a small app to reproduce [here](https://github.com/darkblueorange/elixir_exgboost_bootstrap). 
Sorry it comes with phoenix framework, it has become a reflex for small projects bootstrapping. 

File in dataset example comes from [Kaggle Fraud dataset](https://www.kaggle.com/datasets/mlg-ulb/creditcardfraud). 
It's a bit heavy (160Mo unzipped), it has been uploaded with git lfs. 

Tried different options for training EXGBoost, with no luck. Under 208_601 no crash occurs.

**Comment by darkblueorange** [comment]
Forgot to mention: tested `{:exgboost, "~> 0.2"}` and `{:exgboost, "~> 0.3.1"}`

**Comment by acalejos** [comment]
Thanks for raising the issue and I hopeEXGBoost otherwise has been good to use. I'll take a look tonight. In the meantime, can you include some screenshots with any error messages? Once I know if the error is from the NIF itself it will be easier to troubleshoot.

Thanks!

**Comment by acalejos** [comment]
Ok so Im trying to reproduce and get the following error after pulling your docker compose and running it:
```
iex(1)> MlXgboost.run()
** (RuntimeError) from_csv failed: "Polars Error: Could not parse `oid sha256:76274b691b16a6c49d3f159c883398e03ccd6d1ee12d9d8ee38f4b4b98551a89` as dtype `f64` at column 'Time' (column number 1).\nThe current offset in the file is 43 bytes.\n\nYou might want to try:\n- increasing `infer_schema_length` (e.g. `infer_schema_length=10000`),\n- specifying correct dtype with the `dtypes` argument\n- setting `ignore_errors` to `True`,\n- adding `oid sha256:76274b691b16a6c49d3f159c883398e03ccd6d1ee12d9d8ee38f4b4b98551a89` to the `null_values` list."
    (explorer 0.7.0) lib/explorer/data_frame.ex:598: Explorer.DataFrame.from_csv!/2
    (exgboost_bootstrap 0.1.0) lib/exgboost_bootstrap/xgboost.ex:238: MlXgboost.run/2
```

But I did notice some other things that should be fixed. I realized that I didn't update the training snippet on this page (https://hexdocs.pm/exgboost/EXGBoost.html#module-training) when I updated the params.

You currently have:
```elixir
EXGBoost.train(
      x_train,
      y_train,
      obj: :binary_logistic,
      evals: [{x_test, y_test, "test"}],
      max_depth: 0,
      tree_method: :approx,
      learning_rates: fn i -> i / 100 end,
      num_boost_round: 10,
      early_stopping_rounds: 3,
      params: [max_depth: 3, eval_metric: ["error", "roc", "auc"]]
    )
```

Params no longer are a separate keyword list (that is my fault, I updated the repo README but not on that docs page). You also have `max_depth` set to `0`, which means no depth limit. XGBoost docs states "0 indicates no limit on depth. Beware that XGBoost aggressively consumes memory when training a deep tree." Perhaps try setting that to 3 or 4? 

The new way the params work would be:
```elixir
EXGBoost.train(
      x_train,
      y_train,
      obj: :binary_logistic,
      evals: [{x_test, y_test, "test"}],
      max_depth: 0,
      tree_method: :approx,
      learning_rates: fn i -> i / 100 end,
      num_boost_round: 10,
      early_stopping_rounds: 3,
      max_depth: 3, 
      eval_metric: [:error, :roc, :auc]
    )
```

I'll try a bit more to reproduce

**Comment by acalejos** [comment]
Ok I tried you code and it works past that number:
```
>
iex(3)> MlXgboost.run(208_650)
Iteration 0: %{"test" => %{"rmse" => 0.04178886232117531}}
Iteration 1: %{"test" => %{"rmse" => 0.04147045125340179}}
Iteration 2: %{"test" => %{"rmse" => 0.04083701601001805}}
Iteration 3: %{"test" => %{"rmse" => 0.03991399166991376}}
Iteration 4: %{"test" => %{"rmse" => 0.03876756098980012}}
Iteration 5: %{"test" => %{"rmse" => 0.03744890031752901}}
Iteration 6: %{"test" => %{"rmse" => 0.03593088921803994}}
Iteration 7: %{"test" => %{"rmse" => 0.03433133165395307}}
Iteration 8: %{"test" => %{"rmse" => 0.03266847291178419}}
Iteration 9: %{"test" => %{"rmse" => 0.03098052639752125}}
#Nx.Tensor<
  f32[41730]
  [0.0012042230227962136, 0.001204281230457127, 0.001204253057949245, 0.0012042230227962136, 0.0012042230227962136, 0.0012042230227962136, 0.001204281230457127, 0.0012042515445500612, 0.0012042230227962136, 0.0012042230227962136, 0.0012042230227962136, 0.0012042230227962136, 0.0012042230227962136, 0.0012042230227962136, 0.0012100234162062407, 0.001204281230457127, 0.0012042230227962136, 0.0012042407179251313, 0.0012050310615450144, 0.0012042539892718196, 0.0012042562011629343, 0.001204253057949245, 0.0012042815797030926, 0.0012042230227962136, 0.001204273896291852, 0.0012042230227962136, 0.0012042230227962136, 0.0012042230227962136, 0.0012042230227962136, 0.0012042407179251313, 0.001204281230457127, 0.0012043227907270193, 0.0012042230227962136, 0.0012042539892718196, 0.0012042230227962136, 0.0012042230227962136, 0.0012042230227962136, 0.0012042230227962136, 0.0012042515445500612, 0.0012042230227962136, 0.0012042230227962136, 0.0012042230227962136, 0.001204253057949245, 0.001204281230457127, 0.0012042230227962136, 0.0012043227907270193, 0.0012042230227962136, 0.0012042230227962136, 0.0012042230227962136, 0.0012042230227962136, ...]
>
iex(4)>
```

I'm thinking it might be a resource constraint.  Try setting a smaller max depth. When I trained and output the model, it generated 10 trees, each of max depth ~8 and num_nodes ~= 270. Let me know if that helps. If so, I will need to find out where the error code is getting dropped then.

**Comment by acalejos** [comment]
I haven't dug into the NIF code yet, but according to the XGBoost Documentation (https://xgboost.readthedocs.io/en/latest/tutorials/external_memory.html), 
> When working with large datasets, training XGBoost models can be challenging as the entire dataset needs to be loaded into memory. This can be costly and sometimes infeasible.
> External memory can be used for both training and prediction, but training is the primary use case and it will be our focus in this tutorial. For prediction and evaluation, users can iterate through the data themselves while training requires the full dataset to be loaded into the memory.

So in the meantime, you might be able to chunk your predictions since the training seems to be succeeding. As far as I can tell, the streaming support thats allowed in the Python API which allows users to define an iterator function will not work with a NIF, since there is no concept of a callback function in Elixir-land.

 I can, however, look into a `streaming_train` and `streaming_predict` type functionality which defines an opinionated way of doing it. More to follow. If nothing else, it should error gracefully and let the user know the error better.

**Comment by darkblueorange** [comment]
Wow thanks a lot for all theses hints. 
Sorry for the LFS thing, I should've written it in my README, I just mentioned it on my first post here. 

Considering the resource consumption: I have 64Go RAM, so I would be surprised I run out of it even for that dataset. I have no system issue, no "slowing down" of the rest of the computer for what it's worth. 
Maybe the Erlang process/VM itselft auto-limits itself, and I could allocate more. I didn't look into that yet. I will try again and take some measure, but I'm very new to these exploration things. 

Considering params of EXGBoost.train my bad, but I did try a lot of different combination like at the begining, through livebook. This one was my first attempt actually:
```elixir
EXGBoost.train(x_train, y_train,
     obj: : binary_logistic,
     evals: [{x_train, y_train, "train"}],
     num_boost_rounds: 15
   )
```

I wrote in the @doc of my build_model/3 function (the one calling EXGBoost.train) that I tried with and without max_depth param activated (so going back to default). I also tried with the explicit value to 3 also. 
I actually thought that it might be a memory issue, but didn't know if it was from the Elixir side, the NIF, the C implementation or my machine. So I wanted to allow the algorithm to go to its maximum extent after trying max_depth 3. 

I do some tests again with your hints and I update you.

**Comment by darkblueorange** [comment]
Here is a first update:

My system RAM seems ok: 
<img width="956" alt="Capture d’écran 2023-09-12 à 12 06 58" src="https://github.com/acalejos/exgboost/assets/42878832/208b26f9-347e-4599-9bf8-17929dc0a618">

Still crashing on the exact same number (208601). 


```elixir
EXGBoost.train(
      x_train,
      y_train,
      obj: :binary_logistic,
      evals: [{x_test, y_test, "test"}],
      tree_method: :approx,
      learning_rates: fn i -> i / 100 end,
      num_boost_round: 10,
      early_stopping_rounds: 3,
      max_depth: 3,
      eval_metric: [:error, :auc]
    )
```
(I had to remove :roc because I got: `(ArgumentError) Parameter `eval_metric` must be in [:rmse, :mae, :logloss, :error, :error, :merror, ...], got :roc`

If you can provide me the command to display the model output I could try to generate one on a smaller dataset (as lowering the max_depth didn't succeed for me). Otherwise I try this night a bit late. 

Nota: I'm not in a hurry of anything, I was just trying to discover and give a shot to both this library and XGBoost algorithm. I got your advice of batching/chunking, I could try for the sake of the knowledge of it, but I don't have business imperatives).

**Comment by darkblueorange** [comment]
I tried to read some stuff about `:erlang.memory`, `:instrument.allocations` and finally observer to get some trace of my memory. 
But it's all chinese for me. 
I don't get how the erlang observer can listen to my process. All Erlang things are not understandable for me, I cannot find what I should do in a simple way. 
I cannot find process number, attach the observer to my process etc. 
Sorry ....

**Comment by darkblueorange** [comment]
Ok I finally got to make something work. 
Launching `:observer.start` from iex inside the precise app running. 
Thing is when it crashes everything crashes, including the observer. 
I just have traces when it doesn't. 
Here it is though:

<img width="848" alt="Capture d’écran 2023-09-12 à 22 06 15" src="https://github.com/acalejos/exgboost/assets/42878832/67cd5198-45c3-4520-bff5-2787ff925171">

and I could grasp a view on the process memory footpring (not the "mortal one", but just under):
<img width="922" alt="Capture d’écran 2023-09-12 à 22 06 40" src="https://github.com/acalejos/exgboost/assets/42878832/29cb56b9-86c2-4c43-a355-13fc9a808e88">

**Comment by acalejos** [comment]
Thanks for digging into it a bit more. I might have to wait until the weekend to take a closer look and really get to the bottom of it.  I will post back on this thread when I do.

**Comment by darkblueorange** [comment]
One last addition from my part: 
I succeeded to launch an observer from another process:

1- On the XGBoost process:
`
iex --name server@192.168.1.7 --cookie mycookie -S mix phx.server
`

2- On the observer process: 
`
iex --name observer@192.168.1.7 --cookie mycookie
`
Then still on observer: 
```elixir
Node.connect(:'server@192.168.1.7')
:observer.start
```

Then if :observer.start return :ok, in the graphic interface we can select `Trace overview` panel > `Add Nodes` and select the server process and see it. 

But ... there is a but. When launching our xgboost predict, erlang still crashes, and the observer survives alone. It crashes so fast that we have no traces. Probably because there is no flush just before dying.

**Comment by darkblueorange** [comment]
Doing:
```elixir
iex> Process.flag(:max_heap_size, 102096500)
```
does limit my heap size (to 102Mo I guess?)
and `MlXgboost.run(208600)` succeeds while `MlXgboost.run(208601)` fails (meaning the process is preoperly killed as heap size is reached) with: 
```elixir
[error]      Process:            #PID<0.565.0> on node :"server@192.168.1.7"
     Context:            maximum heap size reached
     Max Heap Size:      102096500
     Total Heap Size:    102106079
     Kill:               true
     Error Logger:       true
     Message Queue Len:  0
     GC Info:            [
  old_heap_block_size: 38323372,
  heap_block_size: 63872288,
  mbuf_size: 0,
  recent_size: 25032822,
  stack_size: 69,
  old_heap_size: 0,
  heap_size: 31846563,
  bin_vheap_size: 5194215,
  bin_vheap_block_size: 2106274,
  bin_old_vheap_size: 0,
  bin_old_vheap_block_size: 46422
]
```
With this heap size my VM crashes every two or three times though. 
I'm surprised with the low limit if it's really 102Mo of memory.

**Comment by darkblueorange** [comment]
So I also tried:
- casting all dataset to integer (maybe such a big dataset in :f64 was indeed not correctly handled)
- predicting on y_test tensor instead of x_test (I believe that was a big mistake from my side, yet prediction on less lines still work with that)

No luck so far.

**Comment by darkblueorange** [comment]
I tried with x_train, x_test, y_train, y_test all filled with 1 thanks to:
```elixir
Nx.broadcast(1, shape)
|> Nx.as_type(:s64)
```
and it does crash at the same exact size.

**Comment by acalejos** [comment]
I'll take another look tonight. Appreciate you poking around more and letting me know your findings

**Comment by acalejos** [comment]
@darkblueorange So after consulting with @seanmor5 he gave some good suggestions which included increasing the stack size allocated to dirty NIF schedulers. So I added the flag `+sssdcpu 128` (ref. https://www.erlang.org/doc/man/erl.html)  to my VM args when launching IEx and it seems to be working. Can you give that a try and let me know if it helps? 

To pass VM args to IEx you can create a file called something like `vm_args.eex` which includes `+sssdcpu 128`, then invoke IEx with `iex --vm-args vm_args.eex -S mix phx.server`

**Comment by acalejos** [comment]
I can now run it on the whole dataset 

```
iex(8)> MlXgboost.run(280000)
Iteration 0: %{"test" => %{"auc" => 0.5, "error" => 0.00178571428571429}}
Iteration 1: %{"test" => %{"auc" => 0.9396040250447227, "error" => 0.00178571428571429}}
Iteration 2: %{"test" => %{"auc" => 0.9396040250447227, "error" => 0.00178571428571429}}
Iteration 3: %{"test" => %{"auc" => 0.9396040250447227, "error" => 0.00178571428571429}}
Training halted at iteration 4
model: %EXGBoost.Booster{ref: #Reference<0.94630759.4035313665.163126>, best_iteration: 4, best_score: 0.9396040250447227}
Entering predict
Calling booster_predict_from_dmatrix
Finished calling booster_predict_from_dmatrix
%EXGBoost.Booster{
  ref: #Reference<0.94630759.4035313665.163126>,
  best_iteration: 4,
  best_score: 0.9396040250447227
}
iex(9)> MlXgboost.run(284808)
Iteration 0: %{"test" => %{"auc" => 0.5, "error" => 0.00165022295565465}}
Iteration 1: %{"test" => %{"auc" => 0.9518895194395681, "error" => 0.00165022295565465}}
Iteration 2: %{"test" => %{"auc" => 0.9519077587664753, "error" => 0.00165022295565465}}
Iteration 3: %{"test" => %{"auc" => 0.9519036432260449, "error" => 0.00165022295565465}}
Training halted at iteration 4
model: %EXGBoost.Booster{ref: #Reference<0.94630759.4035313665.164722>, best_iteration: 4, best_score: 0.9508031103009732}
Entering predict
Calling booster_predict_from_dmatrix
Finished calling booster_predict_from_dmatrix
%EXGBoost.Booster{
  ref: #Reference<0.94630759.4035313665.164722>,
  best_iteration: 4,
  best_score: 0.9508031103009732
}
```

**Comment by darkblueorange** [comment]
It does work !!
I did search in this direction, but was mislead by NobbZ answer: https://elixirforum.com/t/running-elixir-with-limited-heap-size/23633/2?u=darkblueorange

More generally I feel like there was so many different options to play with memory that I couldn't find where/what:
- erlang:system_flag(max_heap_size, Limit) (http://erlang.org/pipermail/erlang-questions/2016-April/088968.html)
- `+MMsco true +Musac false +MMscs 60` (https://elixirforum.com/t/running-elixir-with-limited-heap-size/23633/3)
- set the max heap size of a process: [max_heap_size](https://erlang.org/doc/man/erlang.html#process_flag_max_heap_size) (in this link https://elixirforum.com/t/what-happens-when-erlang-vm-beam-runs-out-of-memory/39386/18)
- https://www.erlang.org/doc/man/erts_alloc.html

etc

Maybe some consensus could be reached in the community, and outdated links in the forum cleaned. 
Thanks a lot anyway

P.S. NIF is considered as a "dirty process" ?

**Comment by darkblueorange** [comment]
Out of curiosity, how did you see that the model generated was 10 trees, of depth 8, with 270 nodes ?
Thanks

**Comment by acalejos** [comment]
> It does work !! I did search in this direction, but was mislead by NobbZ answer: https://elixirforum.com/t/running-elixir-with-limited-heap-size/23633/2?u=darkblueorange
> 
> More generally I feel like there was so many different options to play with memory that I couldn't find where/what:
> 
> * erlang:system_flag(max_heap_size, Limit) (http://erlang.org/pipermail/erlang-questions/2016-April/088968.html)
> * `+MMsco true +Musac false +MMscs 60` (https://elixirforum.com/t/running-elixir-with-limited-heap-size/23633/3)
> * set the max heap size of a process: [max_heap_size](https://erlang.org/doc/man/erlang.html#process_flag_max_heap_size) (in this link https://elixirforum.com/t/what-happens-when-erlang-vm-beam-runs-out-of-memory/39386/18)
> * https://www.erlang.org/doc/man/erts_alloc.html
> 
> etc
> 
> Maybe some consensus could be reached in the community, and outdated links in the forum cleaned. Thanks a lot anyway
> 
> P.S. NIF is considered as a "dirty process" ?

Glad it worked! I will mark this issue closed then but can update the docs to make a note of this potential issue. Yes, NIFs can be marked as needing the dirty scheduler by marking as IO or CPU bound (see https://github.com/acalejos/exgboost/blob/b2c6cadce10ce26c6749fd19c5e34f3d080ea04d/c/exgboost/src/exgboost.c#L70).

Regarding how I knew the shape of the model, you can serialize the booster using `EXGBoost.dump_model/1` and look at the number of nodes, knowing its a binary tree to see the depth. Perhaps adding a better way to see those sorts of characteristics would be nice
